# Hierarchical chain growth
Grow ensembles of disordered biomolecules from fragment libraries

## Overview
Algorithm to assemble full-length chains of disordered proteins / regions from short overlapping fragments [1]. 
As input to perform the hierarchical chain growth (HCG) you need the sequence of the desired IDP and provide a fragment library. 
Fragment libraries are typically generated by molecular dynamics simulations. 

Experimental information on local conformations can be used to improve the quality of ensembles in 
reweighted hierarchical chain growth (RHCG) [2], which have used to study tau K18. 

We provide some example scripts for a truncated version of tau K18 in the examples folder that can be followed to:
- Prepare the fragments to generate the MD fragment libraries using amber tleap.
- Create a MD fragment library by running replica exchange molecular dynamics simulations of the fragments on a SLURM-based cluster.
- Perform HCG or RHCG with a provided MD fragment library and respective weights (for RHCG).

In the examples folder we use a truncated protein sequence of tau K18 to keep file sizes managable. 

## Installation

Example installation in a new conda enviornment 

```bash
conda create -n hcg_py3.7 python=3.7
conda activate hcg_py3.7
git clone https://github.com/bio-phys/hierarchical-chain-growth.git
cd hierarchical-chain-growth
pip install -e . 
```

Python 3.6+ is required. 

## Inputs
- The sequence or topology file of the desired IDP/IDR as .fasta or .pdb, respectively. If you use .fasta make sure the fasta file contains a "header" beginning with ">".
- A MD fragment library. You may prepare the library following the examples given in this repository.
- If you want to perform RHCG: weights for the individual fragments (e.g., from ensemble reweighting against experimental data reporting on _local_ structure properties - such as chemical shifts)

## Examples truncated tauK18
### Prepare fragments
- Use amber tleap to generate (end-capped) fagments of desired length and with desired overlap.
- For "prepare_fragment_list_tleap.py" you need as input a .fasta or .pdb file to get the sequence of the IDP/IDR you want to grow.
- The script outputs a .txt file containint all the fragment sequences that you need to generate using tleap .
	- The output .txt file can be used as input for tleap.
- Here we show the procedure using the sequence of a truncated tauK18 as input (in form of a fasta file)

### REMD simulation of short fragments to generate MD fragment libraries
_[NOTES!!]_
- Example scripts for SLURM-based cluster.
- Need to be adapted to the respective cluster setup and again benchmarked.
- REMD submit scripts for individual fragment - ideally write loop to cover all fragments.
	- Examples shown in run.sh.
	- Commented out lines: pipeline to submit all scripts at once but with dependencies.

### Run chain growth
The files "run_hcg.py" and "run_rhcg.py" show the setup to perform HCG or RHCG for a truncated tau K18. Most of the default arguments are not shown in the script but may as well be adapted, such as the RMSD cut-off, th min RMSD for alignment, or the clash-distance, heavy atoms within this radius are counted as a clash. Have a look at the docs in "hcg_fct.py" for a more detailed explanation.

## Tests
_[TODO!!]_

## References
1 Hierarchical Ensembles of Intrinsically Disordered Proteins at Atomic Resolution in Molecular Dynamics Simulations, 
Lisa M. Pietrek, Lukas S. Stelzl, and Gerhard Hummer,
Journal of Chemical Theory and Computation 2020 16 (1), 725-737, https://pubs.acs.org/doi/abs/10.1021/acs.jctc.9b00809

2 Global Structure of the Intrinsically Disordered Protein Tau Emerges from its Local Structure, 
Lukas S. Stelzl, Lisa M. Pietrek, Andrea Holla, Javier S. Oroz, Mateusz Sikora, Jürgen Köfinger, Benjamin Schuler, Markus Zweckstetter, Gerhard Hummer, 
bioarxiv, http://dx.doi.org/10.1101/2021.11.23.469691
